### ISSUE TITLE
dosbox-x: error while loading shared libraries: libpng15.so.15: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by CHJ85** [joncampbell123/dosbox-x#1826]
Hi there. I am trying to run Dosbox-X on Ubuntu. Had to compile it because the only package I could find was for Red-Hat/Fedora.
Now I can't run it because there's missing library files. Turns out these are not available in any Debian/Ubuntu package. So I had to go look for it in the Fedora repositories. That file did not work. Likely because it wasn't made for Debian based distros.
So please on your next version, either package Dosbox-x for other distros too or use dependencies that are available on every distro. Or better yet, just package it as a Flatpak. That way it'll run on every distro.
Btw. Why would you even package it for Fedora in favor of the Debian base? Kinda weird. I mean, I don't think a lot of gamers use Fedora. They usually stick to Ubuntu, Pop OS or Linux Mint. You know.

**Comment by rderooy** [comment]
@CHJ85 If you successfully compiled it yourself, that means your system must have had the header files for libpng15 installed. Otherwise it would not be looking for libpng15.so.15

For instance, if your system had the header files for libpng16 installed, it would have linked with that and it would be looking for libpng16.so.16

Since you have the header files for libpng15, which I assume you got from your distro repository, you must also be able to install the actual matching libraries from your distro repository.

**Comment by rderooy** [comment]
In the case of Fedora, I have the following two packages installed. the -devel one has the header files to allow me to compile DOSBox-X.

libpng-devel-1.6.37-3.fc32.x86_64
libpng-1.6.37-3.fc32.x86_64

**Comment by CHJ85** [comment]
Yea I figure I didn't compile it correctly. So I kinda just gave up this project before I corrupt my system somehow.
But again, why can't you guys just do flatpak instead?

**Comment by CHJ85** [comment]
Also, I tried the "alien" way of simply converting the RPM to DEB, but that gave me the same error.

**Comment by dreamer** [comment]
@CHJ85 many Linux distros moved away from libpng15 years ago due to security concerns. Some distros (e.g. Fedora) provide it for compatibility reasons, but it's rather rare.

Perhaps DOSBox-X should do it the same way we handle packages in [dosbox-staging](https://github.com/dosbox-staging/dosbox-staging)? We provide simple tarballs built against Ubuntu 18.04 LTS - it's old enough to provide ABI compatibility with most distros, and we encourage Linux packagers to simply create native builds suited for specific distros.

**Comment by rderooy** [comment]
@dreamer Correct, Fedora 32 provides libpng16 as default, and 15 and 12 for backwards compatibility.

@CHJ85 There is no Linux official binary for the latest DOSBox-X release (0.83.5) at the moment. So I'm assuming your using the 0.83.4 RHEL7 release, which defaults to libpng15. Alien cannot fix that, to run that binary you need version 15, and if libpng15 is not available for your distro release it will fail as you noticed.

I would instead focus on fixing your own compiled version. You just need to ensure you have a matching set of the libpng libraries and headers.

**Comment by Wengier** [comment]
I do not use Linux as my primary OS, so I cannot really help this one. @rderooy If possible, perhaps you can try to improve the Linux build script  to make the Linux version easier to handle?

**Comment by CHJ85** [comment]
@Wengier Thanks.

**Comment by CHJ85** [comment]
@dreamer Thank you, I'll give it another go

**Comment by rderooy** [comment]
Here is a "nightly" RPM build for Fedora 32. This should be using libpng16. Although it may have other library issues when run on a different distro.

[dosbox-x-0.83.5-0.fc32.x86_64.zip](https://github.com/joncampbell123/dosbox-x/files/5179473/dosbox-x-0.83.5-0.fc32.x86_64.zip)

**Comment by CHJ85** [comment]
Well. That solve one problem.
But now there's libvga.so.1 missing, which will not work.
I'm telling you. Flatpak is a far better solution because everything is sandboxed.into the same container.

**Comment by grapeli** [comment]
> But now there's libvga.so.1 missing, which will not work.

If someone in 2020 uses the SVGAlib library to build dosbox-x in Fedora 32, it is possible that they have no complete understanding of package building or are deliberately sabotaging.

Fedora has not used SVGAlib in SDL since 2004.
https://src.fedoraproject.org/rpms/SDL/c/bd70e7535bd5f01b7e0148ec74d8cd037e165b29

Hint: add the appropriate modification to the file https://github.com/joncampbell123/dosbox-x/blob/master/vs2015/sdl/build-dosbox.sh

**Comment by dreamer** [comment]
@grapeli - DOSBox-X uses it's own bundled version of SDL1. (I don't know if it is related to using svgalib, but I suspect it might be).

**Comment by grapeli** [comment]
@dreamer But I know that and that's why I mentioned it clearly.
`--enable-video-svga     use SVGAlib video driver [default=yes]`
@rderooy Has it installed svgalib-devel and he joins them forcibly (although I bet that on purpose).

**Comment by rderooy** [comment]
@grapeli did you forget your tinfoil hat?

I don't remember why I had svgalib-devel installed, but when the headers are present, it will link against them. No need to specify any --enable-video-svga.

Here is another RPM without dependency on svgalib.
[dosbox-x-0.83.5-0.fc32.x86_64.zip](https://github.com/joncampbell123/dosbox-x/files/5181532/dosbox-x-0.83.5-0.fc32.x86_64.zip)

**Comment by Wengier** [comment]
@CHJ85 Flatpak support is being prepared, and thanks for the suggestion. In the meantime, you can try the RPM package posted by @rderooy above.

**Comment by rderooy** [comment]
@Wengier your being overly optimistic. I just said that we may possibly implement it, without a timeframe :-)

For now I am still doing some cleanups, moving stuff to contrib

I have an early flatpak yaml file I made yesterday, but it is not yet building successfully.

**Comment by CHJ85** [comment]
@Wengier Thanks. I did try that RPM but there were a few unmet dependencies that wouldn't work, some .so library files. So I'll just wait for the Flatpak. Thanks for your response and thanks for your work on Dosbox-x.

**Comment by rderooy** [comment]
I have created an initial flatpak manifest for DOSBox-X (SDL1)
https://github.com/rderooy/dosbox-x/commit/390ee8c5abf1fd39b25cacf0bd066fd8dba93960

There are some issues, such as no (working) NE2000 support due to setcap needing root permissions to get going, and I don't know how, or even if, this is possible with a flatpak.

flatpak's are meant to be hosted on a website like flathub, which this is obviously not for now. But I think I will try to create a SDL2 version first. And it will need some more testing.

**Comment by CHJ85** [comment]
Thank you so much for your help guys.
I found Dosbox-x on SnapCraft. 
It was updated on September 3rd, so it's being updated.
You might wanna include it in your readme file even though it's not maintained by you guys.

**Comment by Wengier** [comment]
@CHJ85 Thanks for mentioning that! I have added a link to the DOSBox-X SnapCraft page into the documentation (INSTALL page), noting that the Linux packages are built and updated by SnapCraft and not us.

**Comment by CHJ85** [comment]
@rderooy I appreciate that. Thank you. I appreciate that. 
I certainty prefer Flatpak over Snap due to its proprietary backend.

**Comment by rderooy** [comment]
A flatpak version of DOSBox-X 0.83.6 is now available
https://flathub.org/apps/details/com.dosbox_x.DOSBox-X

**Comment by Wengier** [comment]
Great! I have added the Flatpak package to the website and install instruction page for the Linux platform. I believe it is a great improvement for Linux systems!

**Comment by CHJ85** [comment]
@rderooy @Wengier Great work.
Thanks a bunch!

**Comment by CHJ85** [comment]
Two problems though.
The config file is missing in ~/.var/app/com.dosbox_x.DOSBox-X/config/dosbox-x.
And also, there is no sound.

**Comment by rderooy** [comment]
I need to double check sound. I did not try a lot.

As to the config file, this is pretty normal. Also when you install the RPM
or when you install regular dosbox the config file needs to be generated.

Start dosbox-x and run the following command:

config -wcd -all

**Comment by rderooy** [comment]
@CHJ85 

I had a moment to try, and soundblaster sound works here from flatpak.

As to MIDI, unfortunately because of it being a flatpak build, there is no way to access the soundfont under /usr even if I change the flatpak permissions to see the "whole" filesystem, /usr is still filtered and there is no way to access a system installed soundfont.

But you should be able to copy the soundfont in your home directory and then link to that in the config file. Note that the config file location for a flatpak is ~/.var/app/com.dosbox_x.DOSBox-X/config/dosbox-x/dosbox-x-0.83.6.conf
Unfortunately creating a symlink to a soundfont installed in /usr also does not work. Those symlinks are simply invisible from a flatpak.

**Comment by CHJ85** [comment]
@rderooy Thanks. I appreciate your help.
So the only way around this one is to package soundfont with Dosbox-x?
It's a GPL licensed program, so legally that shouldn't be a problem.

**Comment by rderooy** [comment]
@CHJ85 that is something that I had initially considered. flatpak allows "extra" data packages to be downloaded, and I could add a soundfont. However, a decent GM soundfont is rather large, in the order of 500MB large. So I did not feel that was a good idea.

From what I have seen from gnome software center, it will not ask you if you want to download this extra data, and silently download it, while "freezing" the UI making it seem like gnome software center has frozen for what can be a long time.
